src/pluggy/callers.py

```diff
@@ -127,9 +128,17 @@ def _multicall(hook_impls, caller_kwargs, firstresult=False):
         # run all wrapper post-yield blocks
         for gen in reversed(teardowns):
             try:
```

Comment:
WDYT about writing it like this instead?

```py
try:
    gen.send(outcome)
except StopIteration:
    # Exited normally after one yield.
    pass
except BaseException:
    # Raised an exception.
    outcome._excinfo = sys.exc_info()
else:
    # Has a second yield.
    try:
        gen.close()
    except BaseException:
        # Ignored GeneratorExit in second yield, then raised.
        # This is a lost cause; ignore and let the GC handle it.
        pass
    _raise_wrapfail(gen, "has second yield")
```

I added some comments which just repeat the code but help because the code is a bit tricky.

I flattened the try/except because I think it makes the code a bit clearer.

I modified the existing `outcome` instead of creating a new one, because the current code works by passing one mutable `_Result` everywhere and I think it'd be good to preserve this.

Reply:
The only reason of creation of a new `_Result` object is to avoid direct usage of "private" `_excinfo` field. For me it does not matter which variant is used. Just confirm that you believe that it is worth changing to `_excinfo`. Better alternative is adding a method to `_Result` something like `force_current_exception()`. I tried to make minimal changes to avoid arguing on such changes of the interface.

Are you assuming removing of the outer `try`? If so I would object. Exception from the last `raise_wrapfail()`must be trapped into the `outcome` as well. Extra `try` just ensures that there are no holes in the capture exception logic. 

Maybe `KeyboardInterrupt` requires more aggressive treatment with `gen.throw()` to skip teardown code without explicit `finally`. (I assume is that a hook having well behavior does not use `sys.exit()`.)

Reply:
> Just confirm that you believe that it is worth changing to _excinfo. Better alternative is adding a method to _Result something like force_current_exception(). I tried to make minimal changes to avoid arguing on such changes of the interface.

Right it's not great to access a private attribute but since it's internal I don't mind too much.

We can add a `_Result` method for that, but I'd make it private, and then I feel it's not really worth the trouble over just setting `_excinfo` :)

> Are you assuming removing of the outer try? If so I would object. Exception from the last raise_wrapfail()must be trapped into the outcome as well.

Right, I missed that. I still somewhat dislike the nested try for the happy path, so I revise my suggestion to this:

```py
try:
    gen.send(outcome)
    # Has a second yield.
    try:
        gen.close()
    except BaseException:
        # Ignored GeneratorExit in second yield, then raised or yielded (RuntimeError).
        # This is a lost cause; continue and let the GC handle it.
        pass
    _raise_wrapfail(gen, "has second yield")
except StopIteration:
    # Exited normally after one yield.
    pass
except BaseException:
    # Raised an exception.
    outcome._excinfo = sys.exc_info()
```

Reply:
No, I am against ignoring exception raised in response to `gen.close()`.

If `gen.close()` in finally block is really offensive for your, if I have not missed something, it should be
```python
try:
    gen.send(outcome)
    # Second yield is an error, close() is not necessary after StopIteration
    gen.close()
    _raise_wrapfail(gen, "has second yield")
except StopIteration:
    # Exited normally after one yield.
    pass
except BaseException:
    # Raised an exception.
    outcome._excinfo = sys.exc_info()
```
Second reason for nested `try` blocks was likely ensuring `gen.close()` independently of code path. I believed that it was what @goodboy was asking for.

Reply:
OK, that LGTM too. Would just change the comment to

```py
# Has a second yield.
# Try to close it immediately instead of leaving it to the GC.
# Note that If the generator is misbehaved and ignores `GeneratorExit`, the `close()` may raise.
```

> Second reason for nested try blocks was likely ensuring gen.close() independently of code path

I think doing `close` when it's unnecessary obscures the only time it is necessary (second yield), so my preference is to do it only then. @goodboy might still think different though.

Reply:
New `_Result` made from exception however releases result list unlike direct modification of `_excinfo`.